File: SilverFactorial64/Sharith/Factorial/FactorialSquaredDiffProd.cs
// -------- ToujoursEnBeta
// Author & Copyright : Peter Luschny
// License: LGPL version 3.0 or (at your option)
// Creative Commons Attribution-ShareAlike 3.0
// Comments mail to: peter(at)luschny.de
// Created: 2010-03-01

namespace Sharith.Factorial 
{
    using XInt = Arithmetic.XInt;
    using XMath = MathUtils.XMath;

    public class SquaredDiffProd : IFactorialFunction 
    {
        public string Name => "SquaredDiffProduct  ";

        public XInt Factorial(int n)
        {
            if (n < 0)
            {
                throw new System.ArgumentOutOfRangeException(
                          this.Name + ": " + nameof(n) + " >= 0 required, but was " + n);
            }

            if (n < 7)
            {
                return (XInt)(new[] { 1, 1, 2, 6, 24, 120, 720 })[n];
            }

            long h = n / 2;
            var q = h * h;
            var f = new long[(int)h];
            f[0] = (n & 1) == 1 ? 2 * q * n : 2 * q;
            var i = 1;

            for (var d = 1; d < n - 2; d += 2)
            {
                f[i++] = q -= d;
            }

            return XMath.Product(f, f.Length);
        }
    }
} // endOfFactorialSquaredDiffProd

File: JavaFactorial/src/de/luschny/math/factorial/FactorialSquaredDiffProd.java
// Copyright (C) 2004-2009 Peter Luschny, MIT License applies.
// See http://en.wikipedia.org/wiki/MIT_License
// Visit http://www.luschny.de/math/factorial/FastFactorialFunctions.htm
// Comments mail to: peter(at)luschny.de
package de.luschny.math.factorial;

import de.luschny.math.arithmetic.Xint;

public class FactorialSquaredDiffProd implements IFactorialFunction {

    public FactorialSquaredDiffProd() {
    }

    @Override
    public String getName() {
        return "SquaredDiffProduct";
    }

    @Override
    public Xint factorial(int n) {
        if (n < 0) {
            throw new ArithmeticException("Factorial: n has to be >= 0, but was " + n);
        }

        if (n < 4) {
            return Xint.valueOf(n < 2 ? 1 : n == 2 ? 2 : 6);
        }

        long h = n / 2, q = h * h;
        long[] f = new long[(int) h];
        f[0] = (n & 1) == 1 ? 2 * q * n : 2 * q;
        int i = 1;

        for (int d = 1; d < n - 2; d += 2) {
            f[i++] = q -= d;
        }

        return Xint.product(f);
    }
} // endOfFactorialSquaredrDiffProd
